tactics/class_tactics.ml

```diff
@@ -588,6 +614,49 @@ module Search = struct
       search_cut : hints_path;
       search_hints : hint_db; }
 
+  (** This is caching-specific compare, not geneal-purpose *)
+  let compare_autoinfo a b =
+    let (>>==) f c = if f != 0 then f else Lazy.force c in
+    let open Pervasives in
+    compare a.search_dep b.search_dep >>==
+      lazy (compare a.search_only_classes b.search_only_classes) >>==
+      lazy (compare a.search_hints b.search_hints)
+
+  type tc_cache_entry =
+    {
+      (* the following 2 fields are unpacked Goal.t *)
+      tc_cache_goal_sigma : Evd.evar_map;
+      tc_cache_goal_concl : Constr.constr ;
+
+      tc_cache_info: autoinfo;
+      tc_cache_global_hints: hint_db list }
+
+  let tc_cache_entry_cmp a b =
+    let (>>==) f c = if f != 0 then f else Lazy.force c in
+    (* We define our own version of `CErrror.noncritical` to
+          exclude errors like: Anomaly "Universe Coq.Classes.CMorphisms.397 undefined." *)
+    let noncritical x = if CErrors.is_anomaly x then true else CErrors.noncritical x in
```

Comment:
These anomalies always felt really weird but now we are masking them? Why not fix the actual problem which is that these errors should not have been classified as anomalies in the first place?

Reply:
Indeed, this cannot be merged as-is.

Reply:
I am afraid I do not have enough understanding of that is causing these anomalies to fix it myself. Shall I open a ticket?

Reply:
Yes, and ping @mattam82 there.